NIFI-5318 Implement NiFi test harness: initial commit of nifi-testharness

peter-gergely-horvath opened the pull request:
All issues are fixed from the previous pull request comments.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

lfrancke commented on nifi-testharness/src/test/java/org/apache/nifi/testharness/samples/TestUtils.java:
```diff
@@ -0,0 +1,47 @@
+
+    static File getBinaryDistributionZipFile(File binaryDistributionZipDir) {
+
+        File[] files = binaryDistributionZipDir.listFiles((dir, name) ->
```
This can return `null` which makes the next `if` statement fail with a NPE. I think the condition can be extended to

```
if (files == null || files.length == 0)
```

peter-gergely-horvath commented:
Hi @lfrancke I've fixed the issues you noticed. Can you please review the current state?

lfrancke commented:
Thanks! Unfortunately for some reason I can't see your new commit...

(I should maybe also add that I'm not a committer, I wast just interested in this PR)

MikeThomsen commented:
@peter-gergely-horvath can you push that commit?

peter-gergely-horvath commented:
@MikeThomsen The pull request contains the latest changes; after @lfrancke had made his comments, I corrected the issues mentioned, nuked my whole fork and force-pushed my new version.

(I'm wondering what the preferred approach is for you, when there are some suggestions regarding a pull request: maybe it would be worth documenting this in NiFi dev manual?)

MikeThomsen commented:
@peter-gergely-horvath the consensus seems to be just keep pushing new commits and let us do a squashed commit for you because that helps GitHub maintain the context of feedback in a review.

peter-gergely-horvath commented:
@MikeThomsen OK, in the future, I will follow that approach. For this pull request, please review as it is, since the current state contains everything in a squashed commit.

peter-gergely-horvath commented on nifi-testharness/src/main/java/org/apache/nifi/testharness/util/FileUtils.java:
```diff
@@ -0,0 +1,88 @@
+
+    public static void createLink(Path newLink, Path existingFile)  {
+        try {
+            Files.createSymbolicLink(newLink, existingFile);
```
No, I have not tested this on Windows. I would expect it to work properly, since [Windows 10 does support symbolic links](https://blogs.windows.com/buildingapps/2016/12/02/symlinks-windows-10/). (you just need the [correct permission](https://docs.microsoft.com/en-us/windows/security/threat-protection/security-policy-settings/create-symbolic-links))

peter-gergely-horvath commented on nifi-testharness/src/test/java/org/apache/nifi/testharness/samples/Constants.java:
```diff
@@ -0,0 +1,28 @@
+public final class Constants {
+
+    static final File OUTPUT_DIR = new File("./NiFiTest/NiFiReadTest");
+    static final File NIFI_ZIP_DIR = new File("../../nifi-assembly/target");
```
No, these are in the test cases directory: they are there as a *sample* (referred to in the documentation), and will not be packaged to the actual API artifact. 
The paths there refer the NiFi bundle produced during the NiFi build, which seems to be a good default in case you are working on the test harness itself.

MikeThomsen commented on nifi-docs/src/main/asciidoc/developer-guide.adoc:
```diff
@@ -2296,6 +2296,32 @@ threads that should be used to run the Processor can
+to its configuration, including changing processor properties and replacing processor
+classes with mocks.
+
+In order to add the necessary classes to your project,
```
Just to clarify, you mean external projects that use NiFi. To use this with NiFi's code base, a developer would have to add something to the end of the build system that would run the tests after the assembly is built based on what I've seen reviewing this. Is that right or am I missing something?

Just want to make sure in case we have to answer questions from people come in and try to use this to test their contributions.

peter-gergely-horvath replied:
Yes, absolutely. After reading the paragraph once again, I'm wondering if it would make sense to change `us` to `you` (referring to the user), but it is you call, really. 
If you think it makes sense, can you do it while you merge/squash the commit?

peter-gergely-horvath commented:
Hi @MikeThomsen I _intentionally_ have no test configuration in the project (at least for now): they are merely *samples* of what can be done, but they should not be executed as part of the core NiFi build. 
NiFi is a beast, starting and stopping it takes some time, I do not want to add that to the each NiFi build. 

Please create a new Maven project quickstart project (that will have testing configuration enabled) and add the following (replacing `${nifi version}` with the current one) to the dependencies:

```
<dependency>
	<groupId>org.apache.nifi:</groupId>
	<artifactId>nifi-testharness</artifactId>
	<version>${nifi version}</version>
</dependency>
```
once done, you can take the samples into your own project, where you can experiment with the test harness. 
I understand piggybacking on the tests directory is maybe not the perfect place to deliver samples, but given the circumstances I think it is acceptable and could maybe be improved in the future.

peter-gergely-horvath commented:
That makes sense: I'll look into that... once I have a tiny bit of time...

peter-gergely-horvath commented:
Hi @MikeThomsen sorry for the delay, I have been just very busy. It turns out there have been some changes in the flow XML format, which I have fixed against the latest GA version: now all of the tests are executed properly. 

I have also introduced strict version checking: the NiFi version referenced in the flow file is checked against the actual version of the NiFi distribution used so that we can avoid confusion in the future.

If a user tries to run the test harness with a flow created using a different version of NiFi than the one actually used, this will fail fast, preventing a non functioning installation leaving the user being confused why it does not work properly.

MikeThomsen commented:
> Hi @MikeThomsen sorry for the delay, I have been just very busy. 

Thanks. I've got a lot on my plate as well, but will try to find some time early this week to review.

MikeThomsen commented on nifi-testharness/src/test/java/org/apache/nifi/testharness/samples/Constants.java:
```diff
@@ -0,0 +1,32 @@
+
+    // NOTE: you will have to have the NiFi distribution ZIP placed into this directory.
+    // Its version must be the same as the one referenced in the flow.xml, otherwise it will not work!
+    static final File NIFI_ZIP_DIR = new File(System.getProperty("user.home"), "Downloads");
```
Since this is an internal unit test, it should be pointing to the build output of nifi-assembly.

peter-gergely-horvath replied:
This will be relative to the `nifi_testharness_nifi_home` directory, from which the test cases are executed. Since the test case should clear up that directory automatically, I think this should be fine.

MikeThomsen replied:
That folder evaluated to `$HOME/Downloads` for my user account when I ran it. Are you expecting the user to sudo or something?

peter-gergely-horvath commented:
@MikeThomsen This is kind of tricky: NiFi flows refer the precise version of NiFi within the flow XML file, so you have something like this for the processor definitions:
```
      <class>org.apache.nifi.processors.standard.GetHTTP</class>
      <bundle>
        <group>org.apache.nifi</group>
        <artifact>nifi-standard-nar</artifact>
        <version>1.7.1</version>
      </bundle>
```

When you export a flow file to test, I would expect the you to use precisely the same NiFi version as the one used to export the flow file.

I think there is a little bit of confusion here: please note that these tests do not test the test-harness. The test-harness itself is not tested (kind of a chicken-and-egg issue): these test cases are merely _samples_, which are referenced in the documentation: they demonstrate how an _end-user_ could create test cases for his/her own flows.

Please download the ZIP of NiFi 1.7.1 to the `Downloads` directory within your user home (referenced by `org.apache.nifi.testharness.samples.Constants#NIFI_ZIP_DIR` as 
`new File(System.getProperty("user.home"), "Downloads")` ) and try running the test case again.

peter-gergely-horvath commented:
Closed an created a new pull request for the same issue from a clean state: https://github.com/apache/nifi/pull/3165
